tests/unit/test_charm_state.py

```diff
@@ -229,15 +252,55 @@ def test_from_charm_ssh_debug_info():
     )
 
 
-def test_invalid_runner_storage():
+def test_invalid_runner_storage(charm: MagicMock):
     """
     arrange: Setup mocked charm with juju-storage as runner-storage.
     act: Change the runner-storage config to memory.
     assert: Configuration Error raised.
     """
-    charm = MagicMock()
-    charm.model.relations.__getitem__.return_value = [MagicMock()]
     charm.config = {"runner-storage": "not-exist"}
 
-    with pytest.raises(CharmConfigInvalidError):
+    with pytest.raises(CharmConfigInvalidError) as exc:
+        State.from_charm(charm)
+    assert "Invalid runner-storage" in str(exc.value)
+
+
+def test_openstack_config(charm: MagicMock, clouds_yaml: dict):
+    """
+    arrange: Setup mocked charm with openstack-clouds-yaml config.
+    act: Retrieve state from charm.
+    assert: openstack-clouds-yaml config is parsed correctly.
+    """
+    charm.config[charm_state.OPENSTACK_CLOUDS_YAML_CONFIG_NAME] = json.dumps(clouds_yaml)
+    state = State.from_charm(charm)
+    assert state.charm_config.openstack_clouds_yaml == clouds_yaml
+
+
+def test_openstack_config_invalid_yaml(charm: MagicMock):
+    """
+    arrange: Setup mocked charm with openstack-clouds-yaml config containing invalid yaml.
+    act: Retrieve state from charm.
+    assert: CharmConfigInvalidError is raised.
+    """
+    charm.config[charm_state.OPENSTACK_CLOUDS_YAML_CONFIG_NAME] = "invalid_yaml\n-test: test\n"
+
+    with pytest.raises(CharmConfigInvalidError) as exc:
         State.from_charm(charm)
+    assert "Invalid openstack-clouds-yaml config. Invalid yaml." in str(exc.value)
+
+
+def test_openstack_config_invalid_config(charm: MagicMock, clouds_yaml):
+    """
+    arrange: Setup mocked charm with openstack-clouds-yaml and openstack_manager
+     to raise OpenStackInvalidConfigError.
+    act: Retrieve state from charm.
+    assert: CharmConfigInvalidError is raised.
+    """
+    charm.config[charm_state.OPENSTACK_CLOUDS_YAML_CONFIG_NAME] = json.dumps(clouds_yaml)
+    charm_state.openstack_manager.initialize.side_effect = OpenStackInvalidConfigError("invalid")
+
+    with pytest.raises(CharmConfigInvalidError) as exc:
+        State.from_charm(charm)
+    assert "Invalid openstack config. Not able to initialize openstack integration." in str(
+        exc.value
```

Comment:
We probably will need more here as noted earlier

Reply:
Added validation on the `openstack_manager.py` module side of things, I think mocking the behavior here is okay for now. Would type checking be preferred on the `charm_state` side?

Reply:
On a second thought, it makes more sense for the module to take in dict as a parameter than Any. moving this preliminary check to `charm_state.py`